Classes/Service/SessionHandler.php

```diff
@@ -27,7 +36,9 @@ public function restoreCart(string $key): ?Cart
         if (is_string($sessionData)) {
             $cart = unserialize($sessionData);
             if ($cart instanceof Cart) {
-                return $cart;
+                $afterRestoreCartEvent = new AfterRestoreCartEvent($cart);
+                $this->eventDispatcher->dispatch($afterRestoreCartEvent);
+                return $afterRestoreCartEvent->getCart();
```

Comment:
I bet the orignal pointer `$cart` points to the adapted object, doesn't it?

Reply:
Removed assignments after event dispatching.